### ISSUE TITLE
Shouldn't WeakSet.has() accept any type?

### ISSUE THREAD

**Issue by callumlocke** [facebook/flow#5400]
A `WeakSet` is often used as a memo so you can check later if an object has already been encountered (without causing a memory leak). For example, in a function whose job is to cast a given object to a particular type, you might do this:

```js
import MyClass from './MyClass';

const memo: WeakSet<MyClass> = new WeakSet();

export default function castToMyClass(object: MyClass | Object) {
  if (memo.has(object)) return object; // quick return if it's already known

  const instance = new MyClass(object);

  memo.add(instance);

  return instance;
}
```

It's a contrived example, but the point is: the above code fails type checking because when you check `memo.has(object)`, it's not yet known if the passed object is of the type `MyClass`, which is the only type allowed for objects contained in the WeakSet. 

I think it might make more sense for `WeakSet.prototype.has()` to accept `any` type for the passed argument. The `.add()` method should obviously only accept an argument of the correct type for the WeakSet instance, but for `.has()` I think it makes sense to allow any type, since `WeakSet` is specifically for this kind of technique where you might want to verify if an unknown value is in the set already. Providing you're careful about what goes into the set in the first place (`.add()`), there's no harm asking a WeakSet if it contains a value of another type – the result will simply be false.

I don't know if this is possible, but it would be also good for Flow to be able to infer that if a WeakSet's `.has()` returns true for any value, then that value must be of the generic type defined for the WeakSet.

**Comment by SamChou19815** [comment]
No, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol#shared_symbols_in_the_global_symbol_registry
